### ISSUE TITLE
Notifications

### ISSUE THREAD

**Issue by Sneagan** [mirukana/mirage#33]
I've been leaving Riot open in a web window to get notifications even though I'm not using it. I see that Notifications are in the TODO file. Can you expand a bit on your expectations? I might be able to lend a hand with it.

**Comment by mirukana** [comment]
Here's an example of what the `settings.json` could contain:
  
```json
"notificationRules": {
    "default": {
        "mentionsMe": ["bubble", "sound", "highlight"],
        "mentionsRoom": ["bubble", "sound", "highlight"],
        "hasKeywords": ["bubble"],
        "directMessages": ["bubble", "sound", "highlight"],
        "groupMessages": ["bubble"],
        "botMessages": [],
        "roomInvites": ["bubble", "sound", "highlight"],

        "keywords": [],
        "bubbleMsecDuration": -1,
        "avatarsInBubbles": true,
        "messagesInBubbles": true,
        "defaultSound": "default",
        "highlightMsecDuration": -1
    },

    "@some_account:example.org": {
        "groupMessages": ["bubble", "sound"],
        "defaultSound": "/path/to/some/custom/sound.wav"
    },

    "!someRoom:example.org": {
        "keywords": ["issue", "PR", "release"]
    }
}
```

- The default config file should contain the `default` rules dict
- Users can override `default` rules for specific account or rooms (room rules override any account rules and `default`, account rules override `default` but not room rules) 
-  The "what should happen when this occurs" array rules can contain any of the following:
    - `bubble`: show a desktop notification
    - `highlight`: has different effects depending on the user's desktop: sets the urgency hint on X11, flashes the taskbar on Windows, etc
    - `sound`: play the `default_sound`
    - Anything else: consider it's the path to a custom sound (if `sound` is present too, override it) 
- The duration keys can be `-1` for infinite/no timeout or a number of milliseconds

The default config file is defined [here](https://github.com/mirukana/mirage/blob/master/src/backend/user_files.py#L220).

For desktop notifications, it doesn't look like QML has any easy way of making them, but we can make a [`Backend`](https://github.com/mirukana/mirage/blob/master/src/backend/backend.py#L28) function that uses [plyer](https://plyer.readthedocs.io/en/latest/index.html?highlight=notify#plyer.facades.Notification.notify) and is called from [`MatrixClient.register_nio_event`](https://github.com/mirukana/mirage/blob/master/src/backend/matrix_client.py#L1251).

There's the QML [SoundEffect](https://doc.qt.io/qt-5/qml-qtmultimedia-soundeffect.html) type for sounds, but we can worry about those later.

Highlights are already implemented (without rules support) by sending from python an event to QML: see [here](https://github.com/mirukana/mirage/blob/master/src/backend/matrix_client.py#L1308) and [here](https://github.com/mirukana/mirage/blob/master/src/gui/PythonBridge/Privates/EventHandlers.qml#L13).

After all of that is implemented, we can think about GUI controls. Feel free to ask if you have any questions or suggestions.

**Comment by ellietheoracle** [comment]
is there anything stopping the sound part of the notification from being added ? it would make the app alot more usable as it is .

**Comment by mirukana** [comment]
After discussions, I realized the approach I proposed in my first message wouldn't be standard and the right way to do things would be to use the matrix push rules APIs, which is more complicated.

> is there anything stopping the sound part of the notification from being added ? it would make the app alot more usable as it is .

Not really, you could add some code [around here](https://github.com/mirukana/mirage/blob/37a787b7a78676a9dd6874687672351c5060512e/src/backend/matrix_client.py#L1632) to send an event to QML that would be handled in [this file](https://github.com/mirukana/mirage/blob/37a787b7a78676a9dd6874687672351c5060512e/src/gui/PythonBridge/EventHandlers.qml), and would call an [`Audio`](https://doc.qt.io/qt-5/qml-qtmultimedia-audio.html) component's `play()` method.

**Comment by mirukana** [comment]
Push rules and notifications are now implemented in v0.7.0.
